perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -136,14 +143,121 @@ public static boolean authorized(PerunSession sess, String policyDefinition, Lis
+			policyRoles.addAll(policy.getPerunRoles());
+			if (policy.getMfaRules() != null) mfaRules.addAll(policy.getMfaRules());
+		}
 
 		//Fetch super objects like Vo for group etc.
 		Map <String, Set<Integer>> mapOfBeans = fetchAllRelatedObjects(objects);
 
+		if (!mfaAuthorized(sess, mfaRules, mapOfBeans)) {
+			throw new MfaPrivilegeException("Multi-Factor authentication required");
+		}
+
 		return resolveAuthorization(sess, policyRoles, mapOfBeans);
 	}
 
+	/**
+	 * Checks authorization according to MFA rules.
+	 *
+	 * Returns false if there is an MFA rule on an object which is marked as critical,
+	 * and principal is not authorized by MFA and hasn't got a system role.
+	 * If MFA is globally disabled for whole instance, returns true.
+	 *
+	 * @param sess session
+	 * @param mfaRules is a list of maps where each map entry consists from a role name as a key and a role object as a value.
+	 *                    Relation between each map in the list is logical OR and relation between each entry in the map is logical AND.
+	 *                    Example list - (Map1, Map2...)
+	 *                    Example map - key: MFA ; value: Vo
+	 *                                 key: MFA ; value: Group
+	 * @param mapOfBeans is a map of objects against which will be authorization done.
+	 *                    Example map entry - key: Member ; values: (10,15,26)
+	 * @return true if MFA requirements are met, false otherwise
+	 */
+	private static boolean mfaAuthorized(PerunSession sess, List<Map<String, String>> mfaRules, Map<String, Set<Integer>> mapOfBeans) {
+		try {
+			return !BeansUtils.getCoreConfig().isEnforceMfa() || sess.getPerunPrincipal().getRoles().hasRole(Role.MFA)
+						|| hasSystemRole(sess) || !requiresMfa(sess, mfaRules, mapOfBeans);
+		} catch (RoleManagementRulesNotExistsException e) {
+			throw new InternalErrorException("Error checking system roles", e);
+		}
+	}
+
+	/**
+	 * Returns true if at least one of the given MFA rules requires MFA on objects which are marked as critical.
+	 *
+	 * @param sess
+	 * @param mfaRules
+	 * @param mapOfBeans
+	 * @return
+	 */
+	private static boolean requiresMfa(PerunSession sess, List<Map<String, String>> mfaRules, Map<String, Set<Integer>> mapOfBeans) {
+		for (Map<String, String> rule : mfaRules) {
+			// every rule should have exactly one map entry (with key 'MFA')
+			if (!rule.containsKey(Role.MFA)) continue;
+
+			String ruleObject = rule.get(Role.MFA);
+			if (isBlank(ruleObject)) return true;
+
+			Set<Integer> ids = mapOfBeans.get(ruleObject);
+			if (ids != null && ids.stream().anyMatch(id -> isCriticalObject(sess, ruleObject, id))) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Returns true if the object of given type with given ID is marked as critical.
+	 *
+	 * @param sess
+	 * @param objectType
+	 * @param id
+	 * @return
+	 */
+	private static boolean isCriticalObject(PerunSession sess, String objectType, Integer id) {
```

Comment:
This pretty much does the same thing as `isAnyObjectMfaCritical` from AuthzResolver, have you considered using that instead? But I guess then we would need a method that retrieves objects based on bean name and ids, resulting in the same complexity, if not higher. Just a thought.

Reply:
@Johaney-s  beat me to it

Reply:
Yes, I used `isAnyObjectMfaCritical` to save a little bit of code duplication, but as you said, we still need to retrieve the objects.